### DISCUSSION TITLE
Help with ROI activation and stopping a pipeline

### DISCUSSION THREAD

**Discussion by abeWustl** [bonsai-rx/bonsai#1263] [Q&A]
Hi all, 

I’m trying to run an experiment in a shuttle box that requires multiple conditions with tones and shocks that need to be fulfilled (active avoidance-2AA experiment): 

- The mouse is on one side of the chamber
- A tone will play and the length will depend on two conditions (tone will be no longer than 15 seconds):
    - If the mouse moves to the other side then the tone stops and no shock is delivered 
    - The mouse did not move and after 15 seconds the tone will stop and the mouse will be shocked, then if:
        - The mouse moves to the other side, stop the shock
        - The mouse did not move to the other side then continue with the shock (15 seconds max)
- Switch to the other side of chamber and repeat after ITI

The problem that I’m running into is that I can’t figure out how to stop either the tone or shock when the ROI is not active anymore (mouse is not on the side of the chamber where the trial took place). I attached the workflow that I’m working with, but I haven’t been able to figure out the problem with my limited knowledge of Bonsai and some of the examples here in Github and google forums don’t have workflows attached for reference. Thanks!

[GithubActiveAvoidanceV2.zip](https://github.com/bonsai-rx/bonsai/files/10819016/GithubActiveAvoidanceV2.zip)

**Comment by RoboDoig** [comment]
Hi @abeWustl - if I understand correctly, you want to have the initial tone be stopped either if:
a) The mouse moves to the other side of the arena.
b) The mouse doesn't move but 15 seconds elapses.

For the initial tone logic you could have something like this:

![image](https://user-images.githubusercontent.com/11314305/221297134-d31946a9-3a92-4d2c-a61d-4a51b52d6bfe.png)

After the mouse has entered a particular area, the tone is turned on and then the Response node is entered:

![image](https://user-images.githubusercontent.com/11314305/221297299-195c02d2-7b9a-437c-9950-6049431fea0b.png)

Here we merge two streams, one is the 15 second time linked to a 'false' boolean. The second is a subscription to the activity in ROI A, which uses a condition to output 'true' if the activity in ROI is false (i.e. mouse left the ROI). By doing a merge and take on these streams, the Response node will return false if the timer elapses and true if the mouse leaves. In either case the tone is turned off and we branch out using conditions to enter the shock stimulus logic which should be very similar to the tone generation logic.

I've attached my example here, with the ROI activity subjects replaced by key strokes for testing. Hope this helps.

[GithubActiveAvoidanceV2Revised.zip](https://github.com/bonsai-rx/bonsai/files/10828988/GithubActiveAvoidanceV2Revised.zip)
